File: src/ark_recognizer_py/item.py
import json
from dataclasses import dataclass
from typing import Dict

import cv2 as cv
from numpy.typing import NDArray


@dataclass
class Item:
    """
    Represents an item in Arknights. Items have their own item IDs and icons.
    """
    item_id: str
    item_name: str
    item_icon: NDArray

    def __str__(self):
        return "%s(%s)" % (self.item_name, self.item_id)

    def __repr__(self):
        return self.__str__()


def load_item_name_id_dict():
    with open("items/data/item_table.json", "r") as item_data_file:
        item_data = json.load(item_data_file)
    dict_name_id = {}
    for item_obj in item_data['items'].values():
        item_id = item_obj["itemId"]
        item_name = item_obj["name"]
        dict_name_id[item_name] = item_id
    return dict_name_id


ITEM_DICT_NAME_ID: Dict[str, str] = load_item_name_id_dict()


def load_items():
    with open("items/data/item_icon_table.json", "r") as item_data_file:
        item_data = json.load(item_data_file)
    items = []
    for item_obj in item_data:
        item_name = item_obj["name"]
        item_id = ITEM_DICT_NAME_ID[item_name]
        image_id = item_obj["id"]
        image = cv.imread("items/icon/%s.png" % image_id, cv.IMREAD_UNCHANGED)
        items.append(Item(item_id, item_name, image))
    return items


ITEMS: list[Item] = load_items()

File: src/ark_recognizer_cpp/item.cpp
//
// Created by Flandia on 2021/12/11.
//

#include "item.h"
#include <fstream>
#include <nlohmann/json.hpp>
#include <opencv2/imgcodecs.hpp>

using namespace std;
using namespace nlohmann;
using namespace cv;

std::vector<FullItem> load_full_items(istream &in) {
    json json_obj;

    in >> json_obj;

    vector<FullItem> full_items;
    for (auto &item_json_obj: json_obj["items"]) {
        auto item_id = item_json_obj["itemId"].get<string>();
        auto item_name = item_json_obj["name"].get<string>();
        auto sort_id = item_json_obj["sortId"].get<int>();
        FullItem item{item_id, item_name, sort_id};

        full_items.push_back(item);
    }

    return full_items;
}

std::vector<Item> load_items(istream &in, const vector<FullItem> &full_items) {
    json json_obj;

//    ifstream in("ITEMS/data/item_icon_table.json");
    in >> json_obj;

    vector<Item> items;
    for (auto &item_json_obj: json_obj) {
        auto item_id = item_json_obj["id"].get<string>();
        auto item_name = item_json_obj["name"].get<string>();
        auto sb = ostringstream();
        sb << "items/icon/" << item_id << ".png";

        auto full_item = find_if(
                full_items.begin(), full_items.end(),
                [&item_name](const auto &f) { return f.item_name == item_name; }
        );

        auto item_icon = imread(sb.str(), IMREAD_UNCHANGED);
        Item item{full_item->item_id, item_name, item_icon};

        items.push_back(item);
    }

    return items;
}


std::string FullItem::to_string() const {
    std::ostringstream out;
    out << item_name << "(" << item_id << "; FULL)";
    return out.str();
}

std::string Item::to_string() const {
    std::ostringstream out;
    out << item_name << "(" << item_id << ")";
    return out.str();
}

